phpmainthread.go

```diff
@@ -3,108 +3,188 @@ package frankenphp
-		_ = thread.state.requestSafeStateChange(stateShuttingDown)
-		close(thread.drainChan)
-	}
+	mainThread.state.set(stateShuttingDown)
 	close(mainThread.done)
 	for _, thread := range phpThreads {
+		// shut down all reserved threads
+		if thread.state.compareAndSwap(stateReserved, stateDone) {
+			doneWG.Done()
+			continue
+		}
+		// shut down all active threads
 		go func(thread *phpThread) {
-			thread.state.waitFor(stateDone)
-			thread.handlerMu.Unlock()
+			thread.shutdown()
 			doneWG.Done()
 		}(thread)
 	}
+
 	doneWG.Wait()
-	mainThread.state.set(stateShuttingDown)
-	mainThread.state.waitFor(stateDone)
+	mainThread.state.set(stateDone)
+	mainThread.state.waitFor(stateReserved)
 	phpThreads = nil
 }
 
 func (mainThread *phpMainThread) start() error {
 	if C.frankenphp_new_main_thread(C.int(mainThread.numThreads)) != 0 {
 		return MainThreadCreationError
 	}
+
 	mainThread.state.waitFor(stateReady)
 	return nil
 }
 
 func getInactivePHPThread() *phpThread {
+	thread := getPHPThreadAtState(stateInactive)
+	if thread != nil {
+		return thread
+	}
+	thread = getPHPThreadAtState(stateReserved)
+	if thread == nil {
+		return nil
+	}
+	thread.boot()
+	return thread
+}
+
+func getPHPThreadAtState(state stateID) *phpThread {
 	for _, thread := range phpThreads {
-		if thread.state.is(stateInactive) {
+		if thread.state.is(state) {
 			return thread
 		}
 	}
-	panic("not enough threads reserved")
+	return nil
 }
 
 //export go_frankenphp_main_thread_is_ready
 func go_frankenphp_main_thread_is_ready() {
+	mainThread.setAutomaticMaxThreads()
+	if mainThread.maxThreads < mainThread.numThreads {
+		mainThread.maxThreads = mainThread.numThreads
+	}
+
 	mainThread.state.set(stateReady)
-	mainThread.state.waitFor(stateShuttingDown)
+	mainThread.state.waitFor(stateDone)
+}
+
+// max_threads = auto
+// Estimate the amount of threads based on php.ini and system memory_limit
+// If unable to get the system's memory limit, simply double num_threads
+func (mainThread *phpMainThread) setAutomaticMaxThreads() {
+	if mainThread.maxThreads >= 0 {
+		return
+	}
+	perThreadMemoryLimit := int64(C.frankenphp_get_current_memory_limit())
+	totalSysMemory := memory.TotalSysMemory()
+	if perThreadMemoryLimit <= 0 || totalSysMemory == 0 {
+		mainThread.maxThreads = mainThread.numThreads * 2
+		return
+	}
+	maxAllowedThreads := totalSysMemory / uint64(perThreadMemoryLimit)
+	mainThread.maxThreads = int(maxAllowedThreads)
+	logger.Debug("Automatic thread limit", zap.Int("perThreadMemoryLimitMB", int(perThreadMemoryLimit/1024/1024)), zap.Int("maxThreads", mainThread.maxThreads))
 }
 
 //export go_frankenphp_shutdown_main_thread
 func go_frankenphp_shutdown_main_thread() {
-	mainThread.state.set(stateDone)
+	mainThread.state.set(stateReserved)
+}
+
+//export go_get_php_ini_overrides
+func go_get_php_ini_overrides() *C.char {
+	if mainThread.phpIniOverrides == nil {
+		return nil
+	}
+
+	// convert the php.ini overrides from the Caddy config to a C string
```

Comment:
```suggestion
	// convert the php.ini overrides from map to a C string
```

nit: the library should be independent of Caddy